Source/Core/VideoBackends/Metal/MTLUtil.mm

```diff
@@ -0,0 +1,470 @@
+
+// All resources are packed into one descriptor set for compute.
+#define UBO_BINDING(packing, x) layout(packing, set = 0, binding = (x - 1))
+#define SSBO_BINDING(x) layout(std430, set = 2, binding = x)
+#define IMAGE_BINDING(format, x) layout(format, set = 1, binding = x)
+
+// hlsl to glsl function translation
+#define API_METAL 1
+#define float2 vec2
+#define float3 vec3
+#define float4 vec4
+#define uint2 uvec2
+#define uint3 uvec3
+#define uint4 uvec4
+#define int2 ivec2
+#define int3 ivec3
+#define int4 ivec4
+#define frac fract
+#define lerp mix
+)";
+static const std::string_view SUBGROUP_HELPER_HEADER = R"(
+#extension GL_KHR_shader_subgroup_basic : enable
+#extension GL_KHR_shader_subgroup_arithmetic : enable
+#extension GL_KHR_shader_subgroup_ballot : enable
+
+#define SUPPORTS_SUBGROUP_REDUCTION 1
+#define CAN_USE_SUBGROUP_REDUCTION true
+#define IS_HELPER_INVOCATION gl_HelperInvocation
+#define IS_FIRST_ACTIVE_INVOCATION (subgroupElect())
+#define SUBGROUP_MIN(value) value = subgroupMin(value)
+#define SUBGROUP_MAX(value) value = subgroupMax(value)
+)";
+
+static const std::string_view MSL_HEADER =
+    // We know our shader generator leaves unused variables.
+    "#pragma clang diagnostic ignored \"-Wunused-variable\"\n"
+    // These are usually when the compiler doesn't think a switch is exhaustive
+    "#pragma clang diagnostic ignored \"-Wreturn-type\"\n";
+
+static constexpr spirv_cross::MSLResourceBinding
+MakeResourceBinding(spv::ExecutionModel stage, u32 set, u32 binding,  //
+                    u32 msl_buffer, u32 msl_texture, u32 msl_sampler)
+{
+  spirv_cross::MSLResourceBinding resource;
+  resource.stage = stage;
+  resource.desc_set = set;
+  resource.binding = binding;
+  resource.msl_buffer = msl_buffer;
+  resource.msl_texture = msl_texture;
+  resource.msl_sampler = msl_sampler;
+  return resource;
+}
+
+std::string Metal::Util::TranslateShaderToMSL(ShaderStage stage, std::string_view source)
+{
+  std::string full_source;
+
+  std::string_view header = stage == ShaderStage::Compute ? COMPUTE_SHADER_HEADER : SHADER_HEADER;
+  full_source.reserve(header.size() + SUBGROUP_HELPER_HEADER.size() + source.size());
+
+  full_source.append(header);
+  if (Metal::g_features.subgroup_ops)
+    full_source.append(SUBGROUP_HELPER_HEADER);
+  full_source.append(source);
+
+  std::optional<SPIRV::CodeVector> code;
+  switch (stage)
+  {
+  case ShaderStage::Vertex:
+    code = SPIRV::CompileVertexShader(full_source, APIType::Metal, glslang::EShTargetSpv_1_3);
+    break;
+  case ShaderStage::Geometry:
+    PanicAlertFmt("Tried to compile geometry shader for Metal, but Metal doesn't support them!");
+    break;
+  case ShaderStage::Pixel:
+    code = SPIRV::CompileFragmentShader(full_source, APIType::Metal, glslang::EShTargetSpv_1_3);
+    break;
+  case ShaderStage::Compute:
+    code = SPIRV::CompileComputeShader(full_source, APIType::Metal, glslang::EShTargetSpv_1_3);
+    break;
+  }
+  if (!code)
+    return "";
```

Comment:
IMO it would be better to have `TranslateShaderToMSL` return `std::optional<std::string>` rather than an empty string, so that `CreateShaderFromSource`'s check is for if the optional is present rather than for an empty string. (I also personally prefer writing `!code.has_value()` but this is more a style/readability choice; `operator bool` and `has_value` behave identically)